crates/neon/src/instance/mod.rs

```diff
@@ -0,0 +1,72 @@
+use std::any::Any;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use once_cell::sync::OnceCell;
+
+use crate::context::Context;
+use crate::lifecycle::InstanceData;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    COUNTER.fetch_add(1, Ordering::SeqCst)
+}
+
+/// A cell that can be used to allocate data that is global to an instance
+/// of a Neon addon.
+pub struct Global<T> {
+    _type: PhantomData<T>,
+    id: OnceCell<usize>,
+}
+
+impl<T> Global<T> {
+    pub const fn new() -> Self {
+        Self {
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
+    }
+
+    fn id(&self) -> usize {
+        *self.id.get_or_init(next_id)
+    }
+}
+
+impl<T: Any + Send> Global<T> {
+    pub fn borrow<'a, 'b, C>(&self, cx: &'b mut C) -> Option<&'b T>
```

Comment:
Seeing this written out, I think there's value to add a `get_or_init`--even prematurely. Specifically, I'm thinking of the use case where a caller only needs `&T`, but doesn't know if it's initialized yet. This would require `&mut C`. Best case, this requires manually dropping the created item and then re-calling, worst case it requires `unsafe`.

Instead, we could handle that invariant for the caller. I'm okay leaving this for a follow-up PR.

Reply:
I agree, it'll probably be a common case. I think I'll add it. I'm also considering adding `take()`—WDYT?

Reply:
How would that differ from `Option::take`? I.e., `.get_mut(&mut cx).take()`.

Reply:
The difference is that we are giving out `Option<&mut T>` instead of `&mut Option<T>`. This is a valuable API.

Reply:
I think we could minimize the size of our API and give people the full expressiveness of the `Option` API if we boxed the cells of the vec and gave an `&mut Option<T>` out instead. That comes at the cost of extra heap allocation but I think the intended use of globals is to be used sparingly anyway, so that might be a worthwhile tradeoff.

That said, we've been talking about trying an immutable version of this API instead anyway, which would make the question moot.